src/stdio.jl

```diff
@@ -32,20 +32,41 @@ end
 const bufs = Dict{String,IOBuffer}()
 const stream_interval = 0.1
 const max_bytes = 10*1024
+# max output per code cell is 512 kb
+const max_output_per_request = 1 << 19
 
-"""Continually read from (size limited) Libuv/OS buffer into an (effectively unlimited) `IObuffer`
-to avoid problems when the Libuv/OS buffer gets full (https://github.com/JuliaLang/julia/issues/8789).
-Send data immediately when buffer contains more than `max_bytes` bytes. Otherwise, if data is available
-it will be sent every `stream_interval` seconds (see the Timers set up in watch_stdio)"""
+"""
+Continually read from (size limited) Libuv/OS buffer into an `IObuffer` to avoid problems when
+the Libuv/OS buffer gets full (https://github.com/JuliaLang/julia/issues/8789). Send data immediately
+when buffer contains more than `max_bytes` bytes. Otherwise, if data is available it will be sent every
+`stream_interval` seconds (see the Timers set up in watch_stdio). Truncate the output to `max_output_per_request`
+bytes per execution request since excessive output can bring browsers to a grinding halt.
+"""
 function watch_stream(rd::IO, name::AbstractString)
     task_local_storage(:IJulia_task, "read $name task")
     try
         buf = IOBuffer()
         bufs[name] = buf
+        counter = 0
+        curr_msg = ""
         while !eof(rd) # blocks until something is available
             nb = nb_available(rd)
+            parent_id_exists = "msg_id" in keys(execute_msg.header)
+            continuation = curr_msg == execute_msg.header["msg_id"]
+            # if this a new stream then reset the byte counter
+            if parent_id_exists && !continuation
+                curr_msg = execute_msg.header["msg_id"]
+                counter = 0
+            end
             if nb > 0
-                write(buf, read(rd, nb))
+                counter += nb
+                # if this stream as surpassed the maximum output limit then ignore future bytes
+                if parent_id_exists && continuation && counter > max_output_per_request
+                    (counter - nb <= max_output_per_request) && println(orig_STDOUT[], "Excessive output truncated.")
```

Comment:
it is weird to have the `println` as part of the conditional, especially since `println` returns `nothing` ... won't this give an error?  `if true && println("foo"); end` gives an error `TypeError: non-boolean (Void) used in boolean context`

I don't think we want to print a message on every read; I think we just want one warning per cell, and I think we want to print it to the notebook.   If you switch to the global counter as I suggested above, we can just do:
```
if stdio_counter[] > stdio_max[]
   counter[] = 0
   warn("Output > ", stdio_max[], " bytes was truncated.")
end
```
in when the execute completes.